### PR TITLE
[SPARK-37533][SQL] New SQL function: try_element_at

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Add New SQL functions `try_element_at`, which is identical to the `element_at` except that it returns null if error occurs

### Why are the changes needed?

After some data science on real world SQL queries, we found that some users are using the following queries
```
select ... where mapCol.key is not null
select ... where arrayCol[index] is not null
```

### Does this PR introduce _any_ user-facing change?

Yes, an alternative function `try_element_at` which will return null on array index out of bound or map key doesn't exist under ANSI mode.

### How was this patch tested?

Unit tests

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @cloud-fan @entong @allisonwang-db

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -296,7 +296,7 @@ case class Elt(
       val index = indexObj.asInstanceOf[Int]
       if (index <= 0 || index > inputExprs.length) {
         if (failOnError) {
-          throw QueryExecutionErrors.invalidArrayIndexError(index, inputExprs.length)
+          throw QueryExecutionErrors.invalidStringIndexError(index, inputExprs.length)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:299]: the name `StringIndex` is a bit misleading. how about `InputIndex`?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 1 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Sure.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -348,11 +348,13 @@ case class Elt(
       }.mkString)
 
     val indexOutOfBoundBranch = if (failOnError) {
+      // scalastyle:off line.size.limit
       s"""
          |if (!$indexMatched) {
-         |  throw QueryExecutionErrors.invalidArrayIndexError(${index.value}, ${inputExprs.length});
+         |  throw QueryExecutionErrors.invalidStringIndexError(${index.value}, ${inputExprs.length});
```
Comment (allisonwang-db) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:354]: `invalidInputIndexError`

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -82,6 +82,9 @@
     "message" : [ "The fraction of sec must be zero. Valid range is [0, 60]. If necessary set %s to false to bypass this error. " ],
     "sqlState" : "22023"
   },
+  "INVALID_INPUT_INDEX" : {
```
Comment (allisonwang-db) [inline core/src/main/resources/error/error-classes.json:85]: Do we need to add a new error? The error message here seems a little inconsistent between map and array access using `[]` operator:
```
SELECT array(1,2,3)[3];   // Do not hint `try_element_at`
SELECT map('a', 1, 'b', 2)['c'];  // Hint `try_element_at`
```

**Review by allisonwang-db — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks for the review.
Merging to master

**Pull request closed without merging**